src/Features/Core/Portable/AddImport/AbstractAddImportCodeFixProvider.cs

```diff
@@ -288,15 +288,30 @@ private static bool IsHostOrTestWorkspace(Project project)
         /// for the project we found the pe-reference in.
         /// </summary>
         private ImmutableArray<(ProjectId, PortableExecutableReference)> GetUnreferencedMetadataReferences(
-            Project project, HashSet<(ProjectId, PortableExecutableReference)> seenReferences)
+            Project project, HashSet<PortableExecutableReference> seenReferences)
         {
+            var result = ArrayBuilder<(ProjectId, PortableExecutableReference)>.GetInstance();
+
             var solution = project.Solution;
-            return solution.Projects.Where(p => p != project)
-                                    .SelectMany(p => p.MetadataReferences.OfType<PortableExecutableReference>().Select(pe => (p.Id, reference: pe)))
-                                    .Distinct(comparer: this)
-                                    .Where(t => !seenReferences.Contains(t))
-                                    .Where(t => !IsInPackagesDirectory(t.Item2))
-                                    .ToImmutableArray();
+            foreach (var p in solution.Projects)
+            {
+                if (p == project)
+                {
+                    continue;
+                }
+
+                foreach (var reference in p.MetadataReferences)
+                {
+                    if (reference is PortableExecutableReference peReference &&
+                        !IsInPackagesDirectory(peReference) &&
+                        seenReferences.Add(peReference))
```

Comment:
You've lost the `Distinct` and the check against already being in `seenReferences` with this transformation.

If you don't need `Distinct`, do you still need the comparer support at all?